src/metabase/driver/generic_sql/query_processor.clj

```diff
@@ -40,13 +42,40 @@
     [form alias]
     form))
 
+;; TODO - Put this somewhere more generic like QP interface
+(defn- expression-with-name
+  "Return the `Expression` referenced by a given EXPRESSION-NAME."
+  [expression-name]
+  (or (some (fn [expression]
+              (when (= (name (:expression-name expression))
+                       (name expression-name))
+                expression))
+            (:expressions (:query *query*)))
+      (throw (Exception. (format "No expression named '%s'." (name expression-name))))))
 
 (defprotocol ^:private IGenericSQLFormattable
   (formatted [this]
     "Return an appropriate korma form for an object."))
 
 (extend-protocol IGenericSQLFormattable
-  nil (formatted [_] nil)
+  nil    (formatted [_] nil)
+  Number (formatted [this] this)
+  String (formatted [this] this)
+
+  Expression
+  (formatted [{:keys [operator args]}]
+    (apply (case    operator
+             :+     kx/+
+             :-     kx/-
+             :*     kx/*
+             :/     kx//
+             :lower (partial k/sqlfn* :LOWER))
+           (map formatted args)))
+
+  ExpressionRef
+  (formatted [{:keys [expression-name]}]
+    ;; Unfortunately you can't just refer to the expression by name in other clauses like filter, but have to use the original formuala.
+    (formatted (expression-with-name expression-name)))
```

Comment:
One thought here would be to handle this in the query expansion process so that drivers don't have to deal with this.  If you had a query like ...

```
{:filter [">" ["custom-field" "my-field"] 10]
 :expressions {"my-field" ["+" ["field-id" 12] ["field-id" 24]]}}
```

Then the expanded form would come out something like ...

```
{:filter {:filter-type :>
          :field {:expression-name "my-field"
                   :expression {... expanded expression w/ fields ...}}}}
```

Anyways, something like that where the expression definition is getting fully expanded into it's detailed form and then substituted throughout the rest of the query where it's used.